sky/dag.py

```diff
@@ -2,74 +2,253 @@
+    (DAGs) of tasks, representing complex workflows or pipelines.
+
     Examples:
         >>> import sky
-        >>> with sky.Dag() as dag:
-        >>>     task = sky.Task(...)
+        >>> with sky.Dag(name='my_pipeline') as dag:
+        >>>     task1 = sky.Task(name='task1', ...)
+        >>>     task2 = sky.Task(name='task2', ...)
+        >>>     task1 >> task2
     """
 
-    def __init__(self) -> None:
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        tasks: Optional[List['task.Task']] = None,
+        dependencies: Optional[Dict[TaskOrName, Union[List[TaskOrName],
+                                                      TaskOrName]]] = None
+    ) -> None:
+        """Initialize a new DAG.
+
+        Args:
+            name: Optional name for the DAG.
+            tasks: Optional list of Task objects to add to the DAG.
+            dependencies: Optional dictionary specifying task dependencies.
+                          Keys are dependent tasks, values are lists of tasks
+                          they depend on.
+        """
+        self.name = name
         self.tasks: List['task.Task'] = []
-        import networkx as nx  # pylint: disable=import-outside-toplevel
+        self._task_name_lookup: Dict[str, 'task.Task'] = {}
+        self.dependencies: Dict['task.Task', Set['task.Task']] = {}
 
         self.graph = nx.DiGraph()
-        self.name: Optional[str] = None
+
+        # Add tasks
+        if tasks:
+            for task in tasks:
+                self.add(task)
+
+        # Add dependencies
+        if dependencies:
+            for dependent, deps in dependencies.items():
+                self.set_dependencies(dependent, deps)
+
+    def _get_task(self, task_or_name: TaskOrName) -> 'task.Task':
+        """Get a task object from a task or its name.
+
+        Args:
+            task_or_name: Either a Task object or the name of a task.
+
+        Returns:
+            The Task object.
+
+        Raises:
+            ValueError: If the task name is not found in the DAG.
+        """
+        if not isinstance(task_or_name, str):
+            return task_or_name
+        name = task_or_name
+        if name not in self._task_name_lookup:
+            raise ValueError(f'Task {name} not found in DAG')
+        return self._task_name_lookup[name]
 
     def add(self, task: 'task.Task') -> None:
+        """Add a task to the DAG.
+
+        Args:
+            task: The Task object to add.
+
+        Raises:
+            ValueError: If the task already exists in the DAG or if its name
+            is already used.
+        """
+        if task in self.tasks:
+            raise ValueError(f'Task {task.name} already exists in the DAG.')
         self.graph.add_node(task)
         self.tasks.append(task)
+        if task.name is not None:
+            if task.name in self._task_name_lookup:
+                raise ValueError(
+                    f'Task name "{task.name}" is already used in the DAG.')
```

Comment:
We should raise this error before we add the task to `self.graph`?

Reply:
Quite.